### PR TITLE
Merging versioned files into one

### PR DESCRIPTION
**Pull request by pebbe**
Merging versioned files into one, so `go get` will work for all
versions of ZeroMQ without use of -tags, and gozmq will function
identical for all versions of ZeroMQ.

The following stuff is (temporarily) removed:

Zmq 2.1 & 2.2:

```
    RECOVERY_IVL_MSEC = Int64SocketOption(C.ZMQ_RECOVERY_IVL_MSEC)
    SWAP              = Int64SocketOption(C.ZMQ_SWAP)
    MCAST_LOOP        = Int64SocketOption(C.ZMQ_MCAST_LOOP)
    HWM               = UInt64SocketOption(C.ZMQ_HWM)
```

Zmq 2.2:

```
    RCVTIMEO = IntSocketOption(C.ZMQ_RCVTIMEO)
    SNDTIMEO = IntSocketOption(C.ZMQ_SNDTIMEO)
```

Zmq 3.x:

```
    SNDHWM = IntSocketOption(C.ZMQ_SNDHWM)
    RCVHWM = IntSocketOption(C.ZMQ_SNDHWM)

    // TODO Not documented in the man page...
    //LAST_ENDPOINT       = UInt64SocketOption(C.ZMQ_LAST_ENDPOINT)
    FAIL_UNROUTABLE     = BoolSocketOption(C.ZMQ_FAIL_UNROUTABLE)
    TCP_KEEPALIVE       = IntSocketOption(C.ZMQ_TCP_KEEPALIVE)
    TCP_KEEPALIVE_CNT   = IntSocketOption(C.ZMQ_TCP_KEEPALIVE_CNT)
    TCP_KEEPALIVE_IDLE  = IntSocketOption(C.ZMQ_TCP_KEEPALIVE_IDLE)
    TCP_KEEPALIVE_INTVL = IntSocketOption(C.ZMQ_TCP_KEEPALIVE_INTVL)
    // TODO Make this work.
    //TCP_ACCEPT_FILTER   = IntSocketOption(C.ZMQ_TCP_ACCEPT_FILTER)

    // Message options
    MORE = MessageOption(C.ZMQ_MORE)
```

### CODE REVIEW

**Comment by pebbe** [comment]
So, this isn't backward compatible (yet). Perhaps we can add dummy variables for versions that don't support them?

The construct `go get -tags version_id` does not work well. The -tags option is available for get, install and test, but godoc doesn't know about what tags were used to install the package, so the docs will list the wrong variables if ZeroHQ is 2.1 or 3.x. It will also prevent the automatic `go get` by other packages that depend on gozmq.

There is yet another solution. Fork gozmq into gozmq3, develop it further for ZeroMQ version 3 specificly, and remove all version 3 stuff from gozmq. You would need to maintain two separate packages for a while, but in time, gozmq will not be needed anymore. In the mean time, you can further develop gozmq3, adding al the functionality that is new in the ZeroMQ version 3 library.

**Comment by pebbe** [comment]
Dummy variables added in second commit.

Also changed the way socket options are handled, because the type of some changed with version 3. This still leaves it to the user the choose the correct GetSockOpt and SetSockOpt based on their ZeroMQ version.

**Comment by alecthomas** [comment]
Interesting approach, I think I like it. Would you mind merging in master again, there are some conflicts.

File: .gitignore [hunk 1]
```diff
@@ -1,14 +0,0 @@
-*.o
```
Comment (alecthomas) [inline .gitignore:1]: Please undo this deletion.

File: zmq.go [hunk 2]
```diff
@@ -46,15 +100,12 @@ type Socket interface {
...
 
-	SetSockOptInt(option IntSocketOption, value int) error
-	SetSockOptInt64(option Int64SocketOption, value int64) error
-	SetSockOptUInt64(option UInt64SocketOption, value uint64) error
-	SetSockOptString(option StringSocketOption, value string) error
-	GetSockOptInt(option IntSocketOption) (value int, err error)
-	GetSockOptInt64(option Int64SocketOption) (value int64, err error)
-	GetSockOptUInt64(option UInt64SocketOption) (value uint64, err error)
-	GetSockOptString(option StringSocketOption) (value string, err error)
-	GetSockOptBool(option BoolSocketOption) (value bool, err error)
```
Comment (alecthomas) [inline zmq.go:57]: Please preserve all methods if possible. From memory there are bool and "int" (machine word size) options for 2.x. The int options in particular can not be replaced with *Int64 variants on 32-bit architectures.

File: zmq.go [hunk 3]
```diff
@@ -63,15 +114,22 @@ type Socket interface {
...
-type BoolSocketOption int
+type SocketOptionType struct {
+	vartype string
+	option  int
+}
 
 type MessageOption int
 type SendRecvOption int
 
+const (
```
Comment (alecthomas) [inline zmq.go:129]: Perhaps consider using `reflect.Kind` types here.

File: zmq.go [hunk 4]
```diff
@@ -63,15 +114,22 @@ type Socket interface {
 
 type SocketType int
 
-type IntSocketOption int
-type Int64SocketOption int
-type UInt64SocketOption int
-type StringSocketOption int
-type BoolSocketOption int
+type SocketOptionType struct {
```
Comment (alecthomas) [inline zmq.go:121]: I like this approach.

I'm not sure how options invalid on different versions, are handled by zmq. Does it reliably return an error state?

File: zmq.go [hunk 5]
```diff
@@ -46,15 +100,12 @@ type Socket interface {
...
-	GetSockOptInt64(option Int64SocketOption) (value int64, err error)
-	GetSockOptUInt64(option UInt64SocketOption) (value uint64, err error)
-	GetSockOptString(option StringSocketOption) (value string, err error)
-	GetSockOptBool(option BoolSocketOption) (value bool, err error)
+	SetSockOptInt64(option SocketOptionType, value int64) error
+	SetSockOptUInt64(option SocketOptionType, value uint64) error
+	SetSockOptString(option SocketOptionType, value string) error
+	GetSockOptInt64(option SocketOptionType) (value int64, err error)
+	GetSockOptUInt64(option SocketOptionType) (value uint64, err error)
+	GetSockOptString(option SocketOptionType) (value string, err error)
```
Comment (alecthomas) [inline zmq.go:112]: Might be an idea to make these accept `option *SocketOptionType` to avoid a bitwise copy.

File: zmq.go [hunk 6]
```diff
@@ -229,10 +364,20 @@ func (s *zmqSocket) destroy() {
 	}
 }
 
+func optionTypeError(option SocketOptionType, vartype string, prefix string) error {
+	if option.vartype == UNDEF {
+		return errors.New("This option is not available in your current version of ZeroMQ")
```
Comment (alecthomas) [inline zmq.go:384]: Very nice.

**Comment by alecthomas** [comment]
Some preliminary comments.

For testing, I'd suggest running the [examples](https://github.com/imatix/zguide/tree/master/examples/Go) in the zguide. They should all run successfully without modification (modulo the discussed changes to timeouts).

File: README.md [hunk 7]
```diff
@@ -19,19 +19,7 @@ Install gozmq with:
```
Comment (alecthomas) [inline README.md:19]: Please make these changes to .godocdown.md. README.md is generated from that file.

**Comment by pebbe** [comment]
New commit added. It should now merge with latest master.

**Comment by pebbe** [comment]
I don't think this is a valid solution. It solves some problems, but introduces new problems. It is more a survey to see what would be nesessary if you want to keep supporting 0MQ version 2 and 3 in the same library, and want to do away with requiring installation tags to choose the right version.

One problem with the current implementation is that some socket options have wrong argument types voor ZeroMQ version 3. This will lead to run time errors that cannot be fixed in the application using gozmq. You could just solve this by splitting these socket options into `zmq_2_x.go` and `zmq_3_x.go`. This introduces a new problem: application written for gozmq on one machine cannot be compiled on some other machines without the need to fix some code.

My solution was to give all options the same type, with parameter type stored internally. Now you can select the right function to set or get socket options, but you can still choose the wrong one, which will lead to more run time errors. And the only way to find out if you are using the right function is looking at the source code. 

A small part of the solution is to use `int64` in Go for both `int64` and `int` in ZeroMQ, and handle the conversion in gozmq.

But all together, the user is offered a package which in godoc offers no help in handling socket options. Godoc doesn't even show the functions. And there are no compiler errors to help the user choose the right functions for socket options. In this 'pull request' of mine, errors appear later, at run time. That is not a good thing.

My preference is to choose a completely different approach. So, for personal satisfaction, and as a means to learn both ZeroMQ and how to use it in Go, I started my own implementation. See the package documentation here: http://godoc.org/github.com/pebbe/zmq3 (work/play in progress)

**Pull request closed without merging**